app/models/foreman_tasks/task/dynflow_task.rb

```diff
@@ -29,6 +29,15 @@ def cancellable?
     end
 
     def cancel
+      if execution_plan && execution_plan.state == :scheduled
```

Comment:
What would happen if, for some reason the cancelling was not successful? Let's consider that the currently planned task would be actually finishing the run and until we would get to the `execution_plan!.cancel.any?` method, it would already finish, including planning the next round of the action? It might be a corner case, but we should probably count on the worst case scenario here.

It seems that what we are missing in dynflow is ability to hook into the lifecycle of the action to perform some additional logic. I'm thinking of something like

```ruby
class MyAction < Dynflow::Action
  execution_plan_hooks.on_stop do
     # do something when the execution plan stops
  end

  execution_plan_hooks.on_pause do
     # do something when the execution plan got paused (regardless of what step caused the pause)
  end

  execution_plan_hooks.on_success do
     # do something when the execution plan finished successfully
  end

  execution_plan_hooks.on_fail do
     # do something when the execution plan got stopped with failed state (:error, :warning or :cancel)
  end

  on_execution_plan_stopped do
    # do something
  end
end
```

In the first phase, we would probably not need to make this extendable by action, but at least a way to hook into the lifecycle of the execution plan to things like this would be quite beneficial and would allow to do things like this easier that they are right now, where we need to use things like `finalize` for it, which is not always the best place to put the logic, especially when dealing with handling exceptional state.